src/main/java/com/aws/greengrass/clientdevices/auth/iot/Thing.java

```diff
@@ -8,29 +8,143 @@
+ * when the local Thing Registry is updated, or when changes to this Thing are
+ * made in IoT Core. Consider calling the ThingRegistry to retrieve Thing
+ * objects as they are needed rather than storing references long term.
+ */
+@Getter
+public final class Thing implements AttributeProvider {
     public static final String NAMESPACE = "Thing";
     private static final String thingNamePattern = "[a-zA-Z0-9\\-_:]+";
 
-    String thingName;
+    private final int version;
+    private final String thingName;
+    private final List<String> attachedCertificateIds;
+    private boolean modified = false;
 
     /**
-     * Constructor.
+     * Create a new Thing.
+     *
      * @param thingName AWS IoT ThingName
      * @throws IllegalArgumentException If the given ThingName contains illegal characters
      */
-    public Thing(String thingName) {
+    public static Thing of(String thingName) {
+        return Thing.of(0, thingName);
+    }
+
+    /**
+     * Create a new Thing.
+     *
+     * @param version        Thing version
+     * @param thingName      AWS IoT ThingName
+     * @throws IllegalArgumentException If the given ThingName contains illegal characters
+     */
+    public static Thing of(int version, String thingName) {
+        return Thing.of(version, thingName, new ArrayList<>());
+    }
+
+    /**
+     * Create a new Thing.
+     *
+     * @param version        Thing version
+     * @param thingName      AWS IoT ThingName
+     * @param certificateIds Attached certificate IDs
+     * @throws IllegalArgumentException If the given ThingName contains illegal characters
+     */
+    public static Thing of(int version, String thingName, List<String> certificateIds) {
+        if (version < 0) {
+            throw new IllegalArgumentException("Invalid version. Version must not be < 0");
+        }
         if (!Pattern.matches(thingNamePattern, thingName)) {
             throw new IllegalArgumentException("Invalid thing name. The thing name must match \"[a-zA-Z0-9\\-_:]+\".");
         }
+        return new Thing(version, thingName, new ArrayList<>(certificateIds));
+    }
+
+    /**
+     * Create a copy of a Thing.
+     *
+     * @param other Thing to copy
+     */
+    public static Thing of(Thing other) {
+        return Thing.of(other.getVersion(), other.getThingName(), other.getAttachedCertificateIds());
+    }
+
+    /**
+     * Attach a certificate ID.
+     * @param certificateId Certificate ID to attach
+     */
+    public void attachCertificate(String certificateId) {
+        if (attachedCertificateIds.contains(certificateId)) {
+            return;
+        }
+        attachedCertificateIds.add(certificateId);
+        modified = true;
+    }
+
+    /**
+     * Detach a certificate ID.
+     * @param certificateId Certificate ID to detach
+     */
+    public void detachCertificate(String certificateId) {
+        if (!attachedCertificateIds.contains(certificateId)) {
```

Comment:
Why not make `attachedCertificateIds` a `HashSet`?

Reply:
Sure, I guess it will be rare to have too many certIds and we could always use `addIfAbsent`.